### DISCUSSION TITLE
Parallel  SSD1963 vs SPI speed difference using ESP32.

### DISCUSSION THREAD

**Discussion by RudyFiero** [Bodmer/TFT_eSPI#1075] [Q&A - General] [answered]
I am considering using 5" 480x272 displays with a parallel connected SSD1963. Either that or the 3.5" 480x320 SPI LCDs off of Aliexpress.  I will be using the ESP32 as the processor. 

Can anyone give me an idea of the times difference to draw the screen using these options? 

The SPI approach is a lot easier. But if there is a big enough difference I will consider the SSD1963.

**Comment by Bodmer** [comment, marked as answer]
As far as I am aware the SSD1963 ONLY supports a parallel interface but I think you have realised this already.

The time to draw a screen in a real application will depend on what you wish to draw and the time it takes to fetch/decode/prepare any data as well.

SPI displays are generally slower unless you get one that runs at 80MHz or have an application that can effectively use DMA (in which case performance at 40MHz or above is better than 8 bit parallel) for an ESP32. The ESP32 has a 20MHZ I/O clock so can only send 8 bit parallel data at ~40Mbps when allowing for the write strobe and data setup time. STM32 processors can go >2x faster with a parallel display (faster than the display can cope without speed limiting wait periods!)

Old tech displays like the ILI9481 run fine with 8 bit parallel, but only operate at 10MHz SPI rate, so choose carefully.

The common performance test used is the "graphicstest" sketch but this is mainly a function test program, so it has limited value for "real-world" applications. However the "Screen fill" test gives a good idea how long it takes to fill the entire screen 5 times.

These are the results for a SSD1963 800 x 480 screen, so a smaller 272x480 display should be about 3x faster. So you will be able to clear the screen in ~52ms.

```
TFT_eSPI ESP32 SSD1963 800x480 8 bit parallel
Benchmark,                Time (microseconds)
Screen fill (5 times),    769360
Text,                     7473
Lines,                    221595
Horiz/Vert Lines,         72291
Rectangles (outline),     36283
Rectangles (filled),      1893262
Circles (filled),         195700
Circles (outline),        108918
Triangles (outline),      45543
Triangles (filled),       910807
Rounded rects (outline),  63246
Rounded rects (filled),   2888788
Total = 7213266us
Total = 7.2133s
```

I like the ST7796 (but [see here](https://github.com/Bodmer/TFT_eSPI/discussions/898)), which clears the screen in ~62ms:

```
ESP32 ST7796 320x480 40MHz SPI
Benchmark,                Time (microseconds)
Screen fill (5 times),    311693  (16.04 fps)
Text,                     11610
Lines,                    131921
Horiz/Vert Lines,         26807
Rectangles (outline),     14846
Rectangles (filled),      760202
Circles (filled),         91036
Circles (outline),        49494
Triangles (outline),      28627
Triangles (filled),       254510
Rounded rects (outline),  26617
Rounded rects (filled),   763033
Total = 2470396us
Total = 2.4704s
```

**Reply by Vutoan6188** [reply]
Dear sir, I am currently trying to use a 7-inch TFT SSD1963 screen along with your TFT_eSPI library to run your OpenWeatherMap code. Previously, I used a 4-inch ILI9488 TFT screen, and displaying circles or wind directions worked fine. However, now with the 7-inch SSD1963 screen, circles, Moon Icons, or wind directions are distorted. Could you please help me or advise me on how to adjust this?

**Comment by RudyFiero** [comment]
Based on the timings you provided above, it doesn't look like there ia a speed benefit using parallel mode with the ESP32.  But I decided to try it anyway. Unfortunately without success. 

I'm using a board I designed at work. We use a PIC32 in the design. The unit that I am using has sockets for the PIC32 CPU module. Normally we the module soldered in but I have a few that have sockets so that when the programmers blow up the the CPU in their development unit, I can can easily replace it. (we had one programmer who regularly blew things up)

What I did was match the LCD wire connections listed in **Setup50_SSD1963_Parallel.h**. 
// ESP32 pins used
#define TFT_CS   33  // Chip select control pin
#define TFT_DC   15  // Data Command control pin - must use a pin in the range 0-31
#define TFT_RST  32  // Reset pin

#define TFT_WR    4  // Write strobe control pin - must use a pin in the range 0-31
#define TFT_RD    2

#define TFT_D0   12  // Must use pins in the range 0-31 for the data bus
#define TFT_D1   13  // so a single register write sets/clears all bits
#define TFT_D2   26
#define TFT_D3   25
#define TFT_D4   17
#define TFT_D5   16
#define TFT_D6   27
#define TFT_D7   14

I then changed **User_Setup_Select.h**  I commented out
**//#include <User_Setup.h>         // Default setup is root library folder**
and included 
**#include <User_Setups/Setup50_SSD1963_Parallel.h>  // Setup file for ESP32 and SSD1963 TFT display**

I then opened **TFT_Meter_4.INO**, compiled, then loaded to the ESP32. 
What I get is a white screen. I then added some Serial.print statements to see if it gets stuck, and it doesn't. 
I added a Serial.print when the meter deflection counter gets re-initialized to 0, printing out **updateTime**.
All is working, just nothing on the display.

Is there something I missed with the change to the setup?
(I'm a hardware guy and not much of a programmer. Just for fun)

**Comment by RudyFiero** [comment]
One thing I forgot to include above. In **Setup50_SSD1963_Parallel.h**  I changed the display type to
**#define SSD1963_480_DRIVER    // 272 x 480 display**
since that is the display I am using.

**Reply by RudyFiero** [reply]
I figured my problem, mostly. We had first used a PIC32MX series processor, but we ran out of memory. So we switched to the PIC32MZ series. The pinout for the parts are almost the same. One difference is the RD and WR pins, they are moved. And I was using the old part pin locations rather than the new ones. 

So now I get a black screen. Better than just white. But it's late and I should be in bed. Maybe I will figure it out tomorrow.

**Comment by dhionnystrauss** [comment]


**Reply by Basitadam** [reply]
Do you have #define TOUCH_CS  line in your setup file?
